internal/venice-common/src/main/java/com/linkedin/venice/meta/Partition.java

```diff
@@ -30,10 +31,52 @@ public class Partition {
   private final int id;
 
   private final Map<String, List<Instance>> stateToInstancesMap;
+  private final Map<HelixState, List<Instance>> helixStateToInstancesMap;
+  private final Map<ExecutionStatus, List<Instance>> executionStatusToInstancesMap;
+  private final List<Instance> readyInstances;
+  private final List<Instance> workingInstances;
 
   public Partition(int id, Map<String, List<Instance>> stateToInstancesMap) {
     this.id = id;
     this.stateToInstancesMap = stateToInstancesMap;
+
+    // Populate enum maps which are more efficient than the string-based map
+    this.helixStateToInstancesMap = new EnumMap<>(HelixState.class);
+    this.executionStatusToInstancesMap = new EnumMap<>(ExecutionStatus.class);
+    for (Map.Entry<String, List<Instance>> entry: stateToInstancesMap.entrySet()) {
+      try {
+        HelixState helixState = HelixState.valueOf(entry.getKey());
+        this.helixStateToInstancesMap.put(helixState, Collections.unmodifiableList(entry.getValue()));
+      } catch (IllegalArgumentException e) {
```

Comment:
The string-keyed map combines both enums in its key space. So if looking for a string that doesn't exist in one the enums, it will throw an IAE. I can leave a comment for that.

Reply:
I see.
Javadoc should be good enough.
At the first glance, the code is quite confusing.

Reply:
Done.